template/template.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const {promisify} = require('util');
+const recursiveCopy = promisify(require('recursive-copy'));
+
+const mkdir = promisify(fs.mkdir);
+const readdir = promisify(fs.readdir);
+const readFile = promisify(fs.readFile);
+const stat = promisify(fs.stat);
+const writeFile = promisify(fs.writeFile);
+
+module.exports = async templatePath => {
+  const contents = await readFile(templatePath, {encoding: 'UTF-8'});
+
+  const map = {};
+  const parts = contents.split(/<!-- (DOCS \w+) -->/).map((part, index) => {
+    const matcher = /DOCS (\w+)/.exec(part);
+    if (matcher) {
+      part = '';
+      map[matcher[1]] = index;
+    }
+    return part;
+  });
+  const apply = (raw, lang, initialJsState) => {
+    const head = /<head>(.+?)<\/head>/s.exec(raw);
+    if (!head) {
+      throw `Couldn't find head in ${raw}`;
+    }
+    const body = /<body>(.+?)<\/body>/s.exec(raw);
+    if (!body) {
+      throw `Couldn't find body in ${raw}`;
+    }
+
+    const theseParts = parts.slice(0);
+    theseParts[map.PREHEAD] = head[1];
+    theseParts[map.LANG] = `lang="${lang}"`;
+    theseParts[map.BODY] = body[1];
+    theseParts[map.FINAL] = `
+<script type="text/javascript">
+window.initial_state = ${initialJsState}</script>`;
+    return theseParts.join('');
+  };
+  const buildInitialJsStateFromFile = async alternativesSummary => {
+    if (!alternativesSummary) {
+      return "{}";
+    }
+    const readAltSummary = JSON.parse(await readFile(alternativesSummary));
+    return JSON.stringify(module.exports.buildInitialJsState(readAltSummary));
+  };
+  const applyToDir = async (sourcePath, destPath, lang, alternativesSummary, tocMode) => {
+    const initialJsState = await buildInitialJsStateFromFile(alternativesSummary);
+    const entries = await readdir(sourcePath);
+    await mkdir(destPath, {recursive: true});
+    for (var e = 0; e < entries.length; e++) {
```

Comment:
nit: [`Array.prototype.forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) would work here.

Reply:
How does that work with async/await? I think it was upset at me when I tried it.